pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,158 @@
+	defaultTagRgx    *regexp.Regexp
+	timer            *time.Timer
+}
+
+func NewController(kubeClient client.Client, sess *session.Session, instanceProvider *instance.Provider) *Controller {
+	return &Controller{
+		kubeClient:       kubeClient,
+		instanceProvider: instanceProvider,
+		ec2:              ec2.New(sess),
+		defaultTagRgx:    regexp.MustCompile(corev1beta1.NodePoolLabelKey + "/.*"),
+		timer:            time.NewTimer(0),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "ec2-instance-tagger"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	nodeClaim := &corev1beta1.NodeClaim{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, nodeClaim); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !isNodeClaimTaggable(nodeClaim) {
+		return reconcile.Result{}, nil
+	}
+
+	// Extract resource ID from node spec
+	// TODO: Should we retry here? If the ProviderID is malformed is there a case where it is later repaired?
+	resourceID, err := utils.ParseInstanceID(nodeClaim.Status.ProviderID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if instance has already been tagged
+	isTagged, err := c.isInstanceTagged(ctx, resourceID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !isTagged {
+		err = c.tagInstance(nodeClaim, resourceID)
+		if err != nil {
+			return reconcile.Result{Requeue: true}, client.IgnoreNotFound(err)
+		}
+	}
+
+	storedNodeClaim := nodeClaim.DeepCopy()
+	nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged] = "true"
+	err = c.kubeClient.Patch(ctx, nodeClaim, client.MergeFrom(storedNodeClaim))
+	if err != nil {
+		return reconcile.Result{Requeue: true}, client.IgnoreNotFound(err)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				nc := o.(*corev1beta1.NodeClaim)
+				return isNodeClaimTaggable(nc)
+			})),
+	)
+}
+
+func (c *Controller) isInstanceTagged(ctx context.Context, resourceID string) (bool, error) {
+	instance, err := c.instanceProvider.Get(ctx, resourceID)
+	if err != nil {
+		return false, err
+	}
+
+	tag, ok := instance.Tags["Name"]
+	if !ok {
+		return false, nil
+	}
+
+	// Check if the instance has the default tag (karpenter.sh/provisioner-name/*). If so it has either:
+	//   a) Already been tagged by us but it is not reflected on the API server
+	//   b) Tagged by some other process. In this case we will respect that tag.
+	if !c.defaultTagRgx.MatchString(tag) {
+		return true, client.IgnoreNotFound(err)
+	}
+	return false, nil
+}
+
+func (c *Controller) tagInstance(nodeClaim *corev1beta1.NodeClaim, resourceID string) error {
+	// Rate limit to one API call per tag period
+	<-c.timer.C
```

Comment:
Why are we introducing a timer concept here and not using something more native like `time.Sleep`

Reply:
The intention was to ensure that there was at least one second between each call across reconciliations but I suppose since concurrent reconciliations == 1 `time.Sleep` would have pretty much the same effect. It would overall make each reconciliation longer but I don't think there would really be any consequences to that.